\documentclass{article}
\usepackage[round]{natbib}
\usepackage{amsmath,amssymb,amsfonts}%
\usepackage{geometry}%
\usepackage{color}
\usepackage{graphicx}
\usepackage{authblk}
\usepackage{nameref}
\usepackage[right]{lineno}

\begin{document}

\linenumbers
\title{What is an Ancestral Recombination Graph?}
% First authors
\author{Author McAuthory}
% Corresponding

\maketitle

% JK: this is a rough first pass for a slightly different paper. Needs
% substantial revision.
\begin{abstract}
It has recently become possible to infer genetic ancestry in the presence of
recombination at scale for the first time. This has created exciting
possibilities, with many downstream applications becoming possible based on
these inferred genealogies. The inferred genetic ancestries are usually
referred to as Ancestral Recombination Graphs, or ARGs. Although initially well
defined as a graphical representation of the coalescent with recombination
stochastic process, the interpretation has become unclear as it is now
also understood to represent a particular realisation of a genetic
ancestry.
Inference methods do not all infer the same information:
although all output marginal trees along the genome, there is a great deal of
variation in how much information is inferred and retained in the relationships
between the trees. An important programme of work over the coming years is to
develop and refine ancestry inference methods, assessing the relative strengths
and weaknesses of the various approaches. Using the blanket term ARG hides the
important differences in the approaches, and hampers our ability to compare and
improve inference methods. In this paper we discuss the different approaches we
may use to encode genetic ancestry, and of the differing amounts of information
about the ancestral process that we can represent and hope to observe. We
classify existing inference methods according to the type of ancestry they
infer, and discuss the strengths and weaknesses of the inferred structure for
downstream applications.
\end{abstract}

\textbf{Keywords:} Ancestral Recombination Graphs

\section*{Introduction}

% HYW: this is a rough first pass, aiming to be concise
For a long time, the Ancestral Recombination Graph (ARG) has been an object of fascination
to those interested in theoretical evolutionary genetics, as it captures all knowable genetic
history of a collection of individual genomes. Recently advances in our ability to infer ARGs
has led to a resurgence of interest in this field. However, these have also given rise to certain 
terminological confusion, which we aim to clear up in this paper.

One major source of confusion is the use of the term "ARG" to refer to both a backwards-in-time
generative process (Griffiths) and a structure which can be created by that process \citep[e.g.][]{minichiello2006mapping,mathieson2020ancestry}. 

\section*{The ARG as a stochastic process}
The coalescent
process~\citep{kingman1982coalescent,kingman1982genealogy,hudson1983testing,
tajima1983evolutionary}

The Ancestral Recombination Graph (ARG) was introduced by
Griffiths~\citep{griffiths1991two,griffiths1997ancestral}.
The ``big'' ARG~\citep{ethier1990two},
and the ``little ARG'' traversed by
Hudson's algorithm~\citep{hudson1983properties}.

We suggest that the process definition should be called the "coalescent with recombination" (CwR), reserving the term ARG to refer to the graph structure itself.

\section*{The ARG as an encoding of genetic ancestry}

\citep[e.g.][]{minichiello2006mapping,mathieson2020ancestry}.

% HYW: this is a rough first pass. We should take care not to assume that the tree sequence
% approach is the only encoding

This graph *structure*, which can in fact be generated either forwards or backwards in time,
consists of a collection of genomes (represented by *nodes*) connected by paths of genetic
inheritance (*edges*). Since genetic inheritance is unidirectional, the edges are *directed*,
with a "parent" and "child" node. Moreover, there can be no cycles in the graph, therefore the
ARG is technically a form of "directed acyclic graph". One way in this can be ensured is to
associate nodes with specific *times*, and for parent nodes to be older than their children.

% HYW: NB: word carefullyL it should also be possible to encode ARGs even if there is not a single fixed coordinate system

The process of inheritance means that an ARG is a very specific form of DAG. In particular,
although a genome may have more than one parent, because inheritance is particulate,
any particular base pair in a genome only has a *single* parent. Therefore, the paths of
inheritance for a *specific position* in the genome of a set of individuals are described by
a *tree*. Assuming that we can align the genomes in question, we can work our way along
the alignment and by examining the graph we can extract a "local tree" at each position.

To allow this extraction to occur, the graph must be annotated in some way, allowing us to
decide, at any node with more than one parent, which of the parents was the source of
each base pair in the genome. We call any such node a *recombination node*. Fortunately,
the biological details of inheritance mean that this information can be encoded in a succinct
form: we do not have to individually assign each base pair a separate parent. For example,
if we assume that the ARG is describing a single chromosome, and there is only one
recombination per chromosome, we can assign a single *breakpoint* to each recombination
node: base pairs to the left of the breakpoint come from one parent, base pairs to the right
come from the other.

This particular ARG encoding, although common, is somewhat restrictive: not only can
it not be extended to multiple chromosomes or multiple recombinations along a chromosome,
but it cannot easily incorporate the very common process of gene conversion.

% HYW: should discuss the link to biology & mitosis/meiosis: a "node" is strictly a cell - it is shorthand to associate it with an "individual"


\bibliographystyle{plainnat}
\bibliography{paper}

\end{document}
